# Phase 4 Project Notebook
- Author: Jonathan Holt
- Data Science Flex

## Business Problem
- What are the top 5 best zip codes for us to invest in?

## Questions to Answer

## What Models & Metrics I plan on using

## Helper Functions
- Functions provided by Flatiron

In [1]:
def get_datetimes(df):
    """
    Takes a dataframe:
    returns only those column names that can be converted into datetime objects 
    as datetime objects.
    NOTE number of returned columns may not match total number of columns in passed dataframe
    """
    
    return pd.to_datetime(df.columns.values[1:], format='%Y-%m')

In [2]:
def melt_data(df):
    """
    Takes the zillow_data dataset in wide form or a subset of the zillow_dataset.  
    Returns a long-form datetime dataframe 
    with the datetime column names as the index and the values as the 'values' column.
    
    If more than one row is passed in the wide-form dataset, the values column
    will be the mean of the values from the datetime columns in all of the rows.
    """
    
    melted = pd.melt(df, id_vars=['RegionName', 'RegionID', 'SizeRank', 'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

## Loading Data

In [3]:
import pandas as pd

pd.set_option('display.max_rows', 1000) #change the amount of rows displayed

In [4]:
df = pd.read_csv("zillow_data.csv")

In [5]:
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

In [6]:
df.describe()

RegionID    RegionName      SizeRank       1996-04       1996-05  \
count   14723.000000  14723.000000  14723.000000  1.368400e+04  1.368400e+04   
mean    81075.010052  48222.348706   7362.000000  1.182991e+05  1.184190e+05   
std     31934.118525  29359.325439   4250.308342  8.600251e+04  8.615567e+04   
min     58196.000000   1001.000000      1.000000  1.130000e+04  1.150000e+04   
25%     67174.500000  22101.500000   3681.500000  6.880000e+04  6.890000e+04   
50%     78007.000000  46106.000000   7362.000000  9.950000e+04  9.950000e+04   
75%     90920.500000  75205.500000  11042.500000  1.432000e+05  1.433000e+05   
max    753844.000000  99901.000000  14723.000000  3.676700e+06  3.704200e+06   

            1996-06       1996-07       1996-08       1996-09       1996-10  \
count  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04  1.368400e+04   
mean   1.185374e+05  1.186531e+05  1.187803e+05  1.189275e+05  1.191205e+05   
std    8.630923e+04  8.646795e+04  8.665094e+04  8.687208e+04  8.715185e+04   
min    1.160000e+04  1.180000e+04  1.180000e+04  1.200000e+04  1.210000e+04   
25%    6.910000e+04  6.920000e+04  6.937500e+04  6.950000e+04  6.960000e+04   
50%    9.970000e+04  9.970000e+04  9.980000e+04  9.990000e+04  9.995000e+04   
75%    1.432250e+05  1.432250e+05  1.435000e+05  1.437000e+05  1.439000e+05   
max    3.729600e+06  3.754600e+06  3.781800e+06  3.813500e+06  3.849600e+06   

       ...       2017-07       2017-08       2017-09       2017-10  \
count  ...  1.472300e+04  1.472300e+04  1.472300e+04  1.472300e+04   
mean   ...  2.733354e+05  2.748658e+05  2.764646e+05  2.780332e+05   
std    ...  3.603984e+05  3.614678e+05  3.627563e+05  3.644610e+05   
min    ...  1.440000e+04  1.450000e+04  1.470000e+04  1.480000e+04   
25%    ...  1.269000e+05  1.275000e+05  1.282000e+05  1.287000e+05   
50%    ...  1.884000e+05  1.896000e+05  1.905000e+05  1.914000e+05   
75%    ...  3.050000e+05  3.066500e+05  3.085000e+05  3.098000e+05   
max    ...  1.888990e+07  1.870350e+07  1.860530e+07  1.856940e+07   

            2017-11       2017-12       2018-01       2018-02       2018-03  \
count  1.472300e+04  1.472300e+04  1.472300e+04  1.472300e+04  1.472300e+04   
mean   2.795209e+05  2.810953e+05  2.826571e+05  2.843687e+05  2.865114e+05   
std    3.656003e+05  3.670454e+05  3.695727e+05  3.717739e+05  3.724612e+05   
min    1.450000e+04  1.430000e+04  1.410000e+04  1.390000e+04  1.380000e+04   
25%    1.292500e+05  1.299000e+05  1.306000e+05  1.310500e+05  1.319500e+05   
50%    1.925000e+05  1.934000e+05  1.941000e+05  1.950000e+05  1.967000e+05   
75%    3.117000e+05  3.134000e+05  3.151000e+05  3.168500e+05  3.188500e+05   
max    1.842880e+07  1.830710e+07  1.836590e+07  1.853040e+07  1.833770e+07   

            2018-04  
count  1.472300e+04  
mean   2.880399e+05  
std    3.720544e+05  
min    1.380000e+04  
25%    1.324000e+05  
50%    1.981000e+05  
75%    3.211000e+05  
max    1.789490e+07  

[8 rows x 268 columns]

In [7]:
#melt_data(df)

## Checking for Null Values

In [8]:
df.isnull().sum().head(10)

RegionID         0
RegionName       0
City             0
State            0
Metro         1043
CountyName       0
SizeRank         0
1996-04       1039
1996-05       1039
1996-06       1039
dtype: int64

### Analysis

There are many Null Values. 
- For categorical data, 7% of Metro are null (1043 of 14,723).
- Dates from 1996 - mid 2003, also have 7% null values.
- Then it starts to get better. 6% null and decreasing.

What is my decision on null values?
- keep?
- delete?
- change (mean)?

In [9]:
df.reset_index()
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

## Analysis of Categories

In [10]:
#cat_df = df.iloc[:,0:7]
#cat_df.head()

In [11]:
#cat_df.info()

In [12]:
#cat_df['RegionName'] = cat_df['RegionName'].astype('object')
#cat_df.info()

In [13]:
df.nunique()

RegionID      14723
RegionName    14723
City           7554
State            51
Metro           701
CountyName     1212
SizeRank      14723
1996-04        2719
1996-05        2721
1996-06        2744
1996-07        2761
1996-08        2701
1996-09        2740
1996-10        2733
1996-11        2741
1996-12        2735
1997-01        2736
1997-02        2740
1997-03        2758
1997-04        2751
1997-05        2770
1997-06        2772
1997-07        2799
1997-08        2786
1997-09        2759
1997-10        2794
1997-11        2772
1997-12        2795
1998-01        2823
1998-02        2830
1998-03        2834
1998-04        2860
1998-05        2862
1998-06        2857
1998-07        2887
1998-08        2909
1998-09        2912
1998-10        2909
1998-11        2903
1998-12        2945
1999-01        2954
1999-02        2969
1999-03        2993
1999-04        3010
1999-05        3053
1999-06        3047
1999-07        3067
1999-08        3062
1999-09        3082
1999-10        3100


In [14]:
#print("RegionID unique values:", cat_df['RegionID'].nunique())
#print("RegionName unique values:", cat_df['RegionName'].nunique())
#print("City unique values:", cat_df['City'].nunique())
#print("State unique values:", cat_df['State'].nunique())
#print("Metro unique values:", cat_df['Metro'].nunique())
#print("CountyName unique values:", cat_df['CountyName'].nunique())
#print("SizeRank unique values:", cat_df['SizeRank'].nunique())

In [15]:
#cat_df.sort_values(by = ['RegionName'], ascending = True).head()
df.sort_values(by = ['RegionName'], ascending = True).head()

RegionID  RegionName         City State        Metro CountyName  \
5850      58196        1001       Agawam    MA  Springfield    Hampden   
4199      58197        1002      Amherst    MA  Springfield  Hampshire   
11213     58200        1005        Barre    MA    Worcester  Worcester   
6850      58201        1007  Belchertown    MA  Springfield  Hampshire   
14547     58202        1008    Blandford    MA  Springfield    Hampden   

       SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
5850       5851  113100.0  112800.0  112600.0  ...   213900   215700   218200   
4199       4200  161000.0  160100.0  159300.0  ...   333700   334800   336700   
11213     11214  103100.0  103400.0  103600.0  ...   205600   206800   208800   
6850       6851  133400.0  132700.0  132000.0  ...   266100   266300   267000   
14547     14548  117500.0  117300.0  117100.0  ...   202400   202900   205900   

       2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
5850    220100   221100   221700   221700   221700   222700   223600  
4199    338900   340400   342000   344500   347400   350600   353300  
11213   210400   211300   213300   215600   217900   219500   220700  
6850    267500   268000   268100   268100   268800   270000   270600  
14547   208500   207500   205400   204500   206800   210900   214200  

[5 rows x 272 columns]

A google search shows that RegionName is the ZipCode for each Region. However, upon sorting, I discovered that any ZipCode beginning with a 0 was ignoring it and displaying as a 4 digit number. I will use the .str().zfill() method to ensure that all RegionNames are displaying the as 5 digits.

In [16]:
#cat_df['RegionName'] = cat_df['RegionName'].astype(str).str.zfill(5)
df['RegionName'] = df['RegionName'].astype(str).str.zfill(5)

In [17]:
df.sort_values(by = ['RegionName'], ascending = True).head()

RegionID RegionName         City State        Metro CountyName  \
5850      58196      01001       Agawam    MA  Springfield    Hampden   
4199      58197      01002      Amherst    MA  Springfield  Hampshire   
11213     58200      01005        Barre    MA    Worcester  Worcester   
6850      58201      01007  Belchertown    MA  Springfield  Hampshire   
14547     58202      01008    Blandford    MA  Springfield    Hampden   

       SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
5850       5851  113100.0  112800.0  112600.0  ...   213900   215700   218200   
4199       4200  161000.0  160100.0  159300.0  ...   333700   334800   336700   
11213     11214  103100.0  103400.0  103600.0  ...   205600   206800   208800   
6850       6851  133400.0  132700.0  132000.0  ...   266100   266300   267000   
14547     14548  117500.0  117300.0  117100.0  ...   202400   202900   205900   

       2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
5850    220100   221100   221700   221700   221700   222700   223600  
4199    338900   340400   342000   344500   347400   350600   353300  
11213   210400   211300   213300   215600   217900   219500   220700  
6850    267500   268000   268100   268100   268800   270000   270600  
14547   208500   207500   205400   204500   206800   210900   214200  

[5 rows x 272 columns]

In [18]:
#cat_df = cat_df.reset_index()
#cat_df

In [19]:
#cat_df.sort_values(by = ['RegionName'], ascending = True).head()

That seems to have fixed it!

In [20]:
melt_data(df)

value
time                     
1996-04-01  118299.123063
1996-05-01  118419.044139
1996-06-01  118537.423268
1996-07-01  118653.069278
1996-08-01  118780.254312
1996-09-01  118927.528500
1996-10-01  119120.520316
1996-11-01  119345.352236
1996-12-01  119685.084771
1997-01-01  120121.265712
1997-02-01  120544.241450
1997-03-01  120896.623794
1997-04-01  121268.174510
1997-05-01  121655.897398
1997-06-01  122053.105817
1997-07-01  122456.339057
1997-08-01  122879.926927
1997-09-01  123327.402265
1997-10-01  123823.105590
1997-11-01  124353.737669
1997-12-01  125002.601388
1998-01-01  125736.669833
1998-02-01  126450.778111
1998-03-01  127086.257032
1998-04-01  127736.362972
1998-05-01  128385.014978
1998-06-01  129032.702564
1998-07-01  129676.357127
1998-08-01  130324.519617
1998-09-01  130982.618543
1998-10-01  131678.088697
1998-11-01  132399.342442
1998-12-01  133224.702272
1999-01-01  134133.593921
1999-02-01  135020.881128
1999-03-01  135841.586907
1999-04-01  136683.794842
1999-05-01  137532.147293
1999-06-01  138391.809746
1999-07-01  139265.156718
1999-08-01  140160.604954
1999-09-01  141084.832323
1999-10-01  142070.095711
1999-11-01  143104.040330
1999-12-01  144261.328268
2000-01-01  145522.064733
2000-02-01  146772.265654
2000-03-01  147943.683788
2000-04-01  149107.817637
2000-05-01  150247.760649
2000-06-01  151364.608753
2000-07-01  152457.017608
2000-08-01  153530.934463
2000-09-01  154597.187112
2000-10-01  155691.101045
2000-11-01  156804.237598
2000-12-01  157978.351721
2001-01-01  159201.775407
2001-02-01  160410.849711
2001-03-01  161555.950902
2001-04-01  162664.345730
2001-05-01  163717.366844
2001-06-01  164729.619347
2001-07-01  165707.861474
2001-08-01  166666.391466
2001-09-01  167634.309929
2001-10-01  168653.357200
2001-11-01  169714.371301
2001-12-01  170846.292102
2002-01-01  172058.120845
2002-02-01  173311.938336
2002-03-01  174550.062103
2002-04-01  175769.284723
2002-05-01  176957.558267
2002-06-01  178131.760064
2002-07-01  179301.906919
2002-08-01  180477.891430
2002-09-01  181696.493023
2002-10-01  182962.913714
2002-11-01  184453.159933
2002-12-01  186098.173449
2003-01-01  187610.681669
2003-02-01  188930.160006
2003-03-01  190295.616278
2003-04-01  191691.758603
2003-05-01  193112.990429
2003-06-01  194601.804632
2003-07-01  196006.780152
2003-08-01  197610.872872
2003-09-01  199257.667512
2003-10-01  200939.572619
2003-11-01  202586.613546
2003-12-01  204228.359290
2004-01-01  207492.620700
2004-02-01  209244.217982
2004-03-01  211168.039896
2004-04-01  213296.509107
2004-05-01  215615.416305
2004-06-01  218136.246025
2004-07-01  220658.172765
2004-08-01  223353.453128
2004-09-01  226056.895432
2004-10-01  228744.641697
2004-11-01  231299.465974
2004-12-01  233785.465830
2005-01-01  236197.426127
2005-02-01  238563.124596
2005-03-01  240871.017095
2005-04-01  243402.183594
2005-05-01  245934.664560
2005-06-01  248474.874300
2005-07-01  250549.564286
2005-08-01  252925.442857
2005-09-01  255151.628571
2005-10-01  257204.985714
2005-11-01  258995.542857
2005-12-01  260551.178571
2006-01-01  261477.013375
2006-02-01  262687.485771
2006-03-01  263850.576266
2006-04-01  264959.113546
2006-05-01  265878.713717
2006-06-01  266561.959306
2006-07-01  266695.604630
2006-08-01  266818.043031
2006-09-01  266757.139814
2006-10-01  266590.875524
2006-11-01  266316.523468
2006-12-01  266090.974934
2007-01-01  265816.833298
2007-02-01  265624.760689
2007-03-01  265366.333404
2007-04-01  265036.424874
2007-05-01  264491.781890
2007-06-01  263696.341204
2007-07-01  262677.172218
2007-08-01  261586.201276
2007-09-01  260415.577605
2007-10-01  259139.142452
2007-11-01  257695.776045
2007-12-01  256114.812190
2008-01-01  254395.855767
2008-02-01  252515.797676
2008-03-01  250538.729102
2008-04-01  248491.584018
2008-05-01  246309.457353
2008-06-01  243985.661661
2008-07-01  241710.336283
2008-08-01  239648.205310
2008-09-01  237774.272566
2008-10-01  236009.734513
2008-11-01  234330.385841
2008-12-01  2327

## Analysis of Data Values

In [21]:
#df['RegionID_copy'] = df['RegionID']

In [22]:
#df.head()

In [23]:
data_df = df.iloc[:, 7:]
data_df.head()

1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
0  334200.0  335400.0  336500.0  337600.0  338500.0  339500.0  340400.0   
1  235700.0  236900.0  236700.0  235400.0  233300.0  230600.0  227300.0   
2  210400.0  212200.0  212200.0  210700.0  208300.0  205500.0  202500.0   
3  498100.0  500900.0  503100.0  504600.0  505500.0  505700.0  505300.0   
4   77300.0   77300.0   77300.0   77300.0   77400.0   77500.0   77600.0   

    1996-11   1996-12   1997-01  ...  2017-07  2017-08  2017-09  2017-10  \
0  341300.0  342600.0  344400.0  ...  1005500  1007500  1007800  1009600   
1  223400.0  219600.0  215800.0  ...   308000   310000   312500   314100   
2  199800.0  198300.0  197300.0  ...   321000   320600   320200   320400   
3  504200.0  503600.0  503400.0  ...  1289800  1287700  1287400  1291500   
4   77700.0   77700.0   77800.0  ...   119100   119400   120000   120300   

   2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1013300  1018700  1024400  1030700  1033800  1030600  
1   315000   316600   318100   319600   321100   321800  
2   320800   321200   321200   323000   326900   329900  
3  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120500   121000   121500  

[5 rows x 265 columns]

In [ ]:
#data_df.columns = pd.to_datetime(data_df.columns, format='%Y-%m')
#data_df.head()

In [24]:
# changing values to thousands for ease of reading
data_df = data_df.applymap(lambda x: x/1000)
data_df.head()

1996-04  1996-05  1996-06  1996-07  1996-08  1996-09  1996-10  1996-11  \
0    334.2    335.4    336.5    337.6    338.5    339.5    340.4    341.3   
1    235.7    236.9    236.7    235.4    233.3    230.6    227.3    223.4   
2    210.4    212.2    212.2    210.7    208.3    205.5    202.5    199.8   
3    498.1    500.9    503.1    504.6    505.5    505.7    505.3    504.2   
4     77.3     77.3     77.3     77.3     77.4     77.5     77.6     77.7   

   1996-12  1997-01  ...  2017-07  2017-08  2017-09  2017-10  2017-11  \
0    342.6    344.4  ...   1005.5   1007.5   1007.8   1009.6   1013.3   
1    219.6    215.8  ...    308.0    310.0    312.5    314.1    315.0   
2    198.3    197.3  ...    321.0    320.6    320.2    320.4    320.8   
3    503.6    503.4  ...   1289.8   1287.7   1287.4   1291.5   1296.6   
4     77.7     77.8  ...    119.1    119.4    120.0    120.3    120.3   

   2017-12  2018-01  2018-02  2018-03  2018-04  
0   1018.7   1024.4   1030.7   1033.8   1030.6  
1    316.6    318.1    319.6    321.1    321.8  
2    321.2    321.2    323.0    326.9    329.9  
3   1299.0   1302.7   1306.4   1308.5   1307.0  
4    120.3    120.3    120.5    121.0    121.5  

[5 rows x 265 columns]

In [25]:
data_df.describe().round(2)

1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
count  13684.0  13684.00  13684.00  13684.00  13684.00  13684.00  13684.00   
mean     118.3    118.42    118.54    118.65    118.78    118.93    119.12   
std       86.0     86.16     86.31     86.47     86.65     86.87     87.15   
min       11.3     11.50     11.60     11.80     11.80     12.00     12.10   
25%       68.8     68.90     69.10     69.20     69.38     69.50     69.60   
50%       99.5     99.50     99.70     99.70     99.80     99.90     99.95   
75%      143.2    143.30    143.22    143.22    143.50    143.70    143.90   
max     3676.7   3704.20   3729.60   3754.60   3781.80   3813.50   3849.60   

        1996-11   1996-12   1997-01  ...   2017-07   2017-08   2017-09  \
count  13684.00  13684.00  13684.00  ...  14723.00  14723.00  14723.00   
mean     119.35    119.69    120.12  ...    273.34    274.87    276.46   
std       87.48     87.91     88.41  ...    360.40    361.47    362.76   
min       12.20     12.30     12.50  ...     14.40     14.50     14.70   
25%       69.80     70.00     70.30  ...    126.90    127.50    128.20   
50%      100.10    100.30    100.60  ...    188.40    189.60    190.50   
75%      144.12    144.30    144.60  ...    305.00    306.65    308.50   
max     3888.90   3928.80   3964.60  ...  18889.90  18703.50  18605.30   

        2017-10   2017-11   2017-12   2018-01   2018-02   2018-03   2018-04  
count  14723.00  14723.00  14723.00  14723.00  14723.00  14723.00  14723.00  
mean     278.03    279.52    281.10    282.66    284.37    286.51    288.04  
std      364.46    365.60    367.05    369.57    371.77    372.46    372.05  
min       14.80     14.50     14.30     14.10     13.90     13.80     13.80  
25%      128.70    129.25    129.90    130.60    131.05    131.95    132.40  
50%      191.40    192.50    193.40    194.10    195.00    196.70    198.10  
75%      309.80    311.70    313.40    315.10    316.85    318.85    321.10  
max    18569.40  18428.80  18307.10  18365.90  18530.40  18337.70  17894.90  

[8 rows x 265 columns]

## Slicing out Years - data_df

In [27]:
yr_1996 = data_df.iloc[:, :9]
yr_1996.head()

1996-04  1996-05  1996-06  1996-07  1996-08  1996-09  1996-10  1996-11  \
0    334.2    335.4    336.5    337.6    338.5    339.5    340.4    341.3   
1    235.7    236.9    236.7    235.4    233.3    230.6    227.3    223.4   
2    210.4    212.2    212.2    210.7    208.3    205.5    202.5    199.8   
3    498.1    500.9    503.1    504.6    505.5    505.7    505.3    504.2   
4     77.3     77.3     77.3     77.3     77.4     77.5     77.6     77.7   

   1996-12  
0    342.6  
1    219.6  
2    198.3  
3    503.6  
4     77.7

In [30]:
yr_1997 = data_df.iloc[:, 9:21]
yr_1998 = data_df.iloc[:, 21:33]
yr_1999 = data_df.iloc[:, 33:45]

## CONTINUE TO SLICE OUT THE YEARS

## Seeing what happens when i melt just one month

In [35]:
df.head(2)

RegionID RegionName      City State              Metro CountyName  \
0     84654      60657   Chicago    IL            Chicago       Cook   
1     90668      75070  McKinney    TX  Dallas-Fort Worth     Collin   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  

[2 rows x 272 columns]

In [39]:
melt_test = df.iloc[:, :8]
melt_test.head()

RegionID RegionName      City State              Metro CountyName  \
0     84654      60657   Chicago    IL            Chicago       Cook   
1     90668      75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982      77494      Katy    TX            Houston     Harris   
3     84616      60614   Chicago    IL            Chicago       Cook   
4     93144      79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04  
0         1  334200.0  
1         2  235700.0  
2         3  210400.0  
3         4  498100.0  
4         5   77300.0

In [40]:
melt_data(melt_test)

value
time                     
1996-04-01  118299.123063

# Merge Attempt

In [ ]:
#data_df = data_df.reset_index()

In [ ]:
#data_df

In [ ]:
#merged_df = cat_df.merge(data_df, on='index')

In [ ]:
#merged_df

## Slicing out Years and attempting to Melt - Full DF

In [ ]:
df.head()

In [ ]:
df.iloc[:, :16].head(2)

In [ ]:
yr_1996 = df.iloc[:, :16]
yr_1996.head()

In [ ]:
melt_data(yr_1996)

# Archive

In [ ]:
#data_df.isnull().sum()

In [ ]:
#df.columns[7:]

In [ ]:
#date_time_cols = pd.to_datetime(df.columns[7:])
#date_time_cols

In [ ]:
#cat_cols = df.columns[:7]
#cat_cols

In [ ]:
#new_cols = cat_cols + date_time_cols
#new_cols

In [ ]:
#df.rename(columns[7:] = date_time_cols)

## Transposing DF

In [ ]:
flipped_df = df.transpose()

In [ ]:
flipped_df.head(10)

In [ ]:
#melt_data(flipped_df)
melt_data(df)

In [ ]:
#roi = table9.apply(lambda x: x['total_profit'] / x['production_budget'], axis=1)
test = df.loc[8]
#test

In [ ]:
melt_data(test)

## Exploring SizeRank

In [ ]:
df.sort_values('SizeRank').head()

In [ ]:
df.sort_values(by = ['State', 'City'], ascending = True)

In [ ]:
cat_data = df[['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank']]

In [ ]:
cat_data

In [ ]:
df.head()

In [ ]:
date_data = df.drop(['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank'], axis=1)

In [ ]:
date_data.head()

In [ ]:
date_time_data = get_datetimes(date_data)

In [ ]:
date_time_data

In [ ]:
len(date_time_data)

In [ ]:
list(date_data.columns)

In [ ]:
#df2 = date_data.append(date_time_data)

## Changing column names to DateTime Format

In [ ]:
#df.columns = pd.to_datetime(df.columns)
date_data.columns = pd.to_datetime(date_data.columns)

In [ ]:
date_data

### Merging with Categories

In [ ]:
#imdb_df = pd.merge(table7, table4, on= 'tconst', how='inner')
base_df = pd.merge(cat_data, date_data, how='outer')

# Mod 4 Project - Starter Notebook

This notebook has been provided to you so that you can make use of the following starter code to help with the trickier parts of preprocessing the Zillow dataset. 

The notebook contains a rough outline the general order you'll likely want to take in this project. You'll notice that most of the areas are left blank. This is so that it's more obvious exactly when you should make use of the starter code provided for preprocessing. 

**_NOTE:_** The number of empty cells are not meant to infer how much or how little code should be involved in any given step--we've just provided a few for your convenience. Add, delete, and change things around in this notebook as needed!

## Some Notes Before Starting

This project will be one of the more challenging projects you complete in this program. This is because working with Time Series data is a bit different than working with regular datasets. In order to make this a bit less frustrating and help you understand what you need to do (and when you need to do it), we'll quickly review the dataset formats that you'll encounter in this project. 

### Wide Format vs Long Format

If you take a look at the format of the data in `zillow_data.csv`, you'll notice that the actual Time Series values are stored as separate columns. Here's a sample: 

<img src='https://raw.githubusercontent.com/learn-co-students/dsc-mod-4-project-seattle-ds-102819/master/images/df_head.png'>

You'll notice that the first seven columns look like any other dataset you're used to working with. However, column 8 refers to the median housing sales values for April 1996, column 9 for May 1996, and so on. This This is called **_Wide Format_**, and it makes the dataframe intuitive and easy to read. However, there are problems with this format when it comes to actually learning from the data, because the data only makes sense if you know the name of the column that the data can be found it. Since column names are metadata, our algorithms will miss out on what dates each value is for. This means that before we pass this data to our ARIMA model, we'll need to reshape our dataset to **_Long Format_**. Reshaped into long format, the dataframe above would now look like:

<img src='https://raw.githubusercontent.com/learn-co-students/dsc-mod-4-project-seattle-ds-102819/master/images/melted1.png'>

There are now many more rows in this dataset--one for each unique time and zipcode combination in the data! Once our dataset is in this format, we'll be able to train an ARIMA model on it. The method used to convert from Wide to Long is `pd.melt()`, and it is common to refer to our dataset as 'melted' after the transition to denote that it is in long format. 

## Helper Functions Provided

Melting a dataset can be tricky if you've never done it before, so you'll see that we have provided a sample function, `melt_data()`, to help you with this step below. Also provided is:

* `get_datetimes()`, a function to deal with converting the column values for datetimes as a pandas series of datetime objects
* Some good parameters for matplotlib to help make your visualizations more readable. 

Good luck!


## Step 1: Load the Data/Filtering for Chosen Zipcodes

## Step 2: Data Preprocessing

In [ ]:
def get_datetimes(df):
    """
    Takes a dataframe:
    returns only those column names that can be converted into datetime objects 
    as datetime objects.
    NOTE number of returned columns may not match total number of columns in passed dataframe
    """
    
    return pd.to_datetime(df.columns.values[1:], format='%Y-%m')

## Step 3: EDA and Visualization

In [ ]:
font = {'family' : 'normal',
        'weight' : 'bold',
        'size'   : 22}

matplotlib.rc('font', **font)

# NOTE: if you visualizations are too cluttered to read, try calling 'plt.gcf().autofmt_xdate()'!

## Step 4: Reshape from Wide to Long Format

In [ ]:
def melt_data(df):
    """
    Takes the zillow_data dataset in wide form or a subset of the zillow_dataset.  
    Returns a long-form datetime dataframe 
    with the datetime column names as the index and the values as the 'values' column.
    
    If more than one row is passes in the wide-form dataset, the values column
    will be the mean of the values from the datetime columns in all of the rows.
    """
    
    melted = pd.melt(df, id_vars=['RegionName', 'RegionID', 'SizeRank', 'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

## Step 5: ARIMA Modeling

## Step 6: Interpreting Results